doc/policy/version3_transactions.md

```diff
@@ -0,0 +1,16 @@
+# Transactions with nVersion 3
+
+A transaction with its `nVersion` field set to 3 ("V3 transactions") are allowed in mempool and
+transaction relay.
+
+The following set of policies apply to V3 transactions:
+
+1. Any descendant of an unconfirmed V3 transaction must also be V3.
+
+2. A V3 transaction can be replaced by another V3 transaction, even if it does not signal BIP125
+   replaceability. Use the (`-mempoolfullrbf`) configuration option to allow transaction
+   replacement without enforcement of any opt-in signaling rule.
+
+3. A V3 transaction cannot have more than 1 descendant.
+
+4. A V3 transaction that has a V3 ancestor cannot be larger than 4000 virtual bytes.
```

Comment:
I believe this falls short in that the parent V3 transaction can be just south of 101kvB, even in ANYONECANPAY and SIGHASH_SINGLE like scenarios where additional inputs cannot be excluded by the presigning parties.

In imprecise language, this is what I expect to be required to solve it: "When evaluating a V3 transaction, if its total mempool package size after entry into the mempool would exceed 4000 virtual bytes, it is rejected."

Reply:
Happy to change this of course, but just want to clarify what you're suggesting.

> In imprecise language, this is what I expect to be required to solve it: "When evaluating a V3 transaction, if its total mempool package size after entry into the mempool would exceed 4000 virtual bytes, it is rejected."

I interpret this to mean that we should change both ancestor and descendant limits to 4000vB. This also means an individual V3 tx is not allowed to be more than 4000vB by itself. So in the LN case, a commitment tx isn't allowed to be more than 4000vB. I believe this would necessitate lowering the max in-flight HTLC limit. Is that accurate and if so, is it ok?

Reply:
Your comment seems correct. Will have to think about this some more with respect to the case today.

Reply:
> I believe this falls short in that the parent V3 transaction can be just south of 101kvB,

Correct me if I'm wrong, but I think this case would be something like: you're trying to confirm tx A which pays no fees itself; your counterparty publishes a package of tx A, B and C; C spends both A and B; A and C are both small, but B is 99kvB with an ephemeral output that is spent by C. You can't replace C without also replacing B because of the ephemeral output.

What you could do, if the ephemeral output is anyone can spend, is replace C in two steps: one that creates tx D that spends B's ephemeral output (it's anyone can spend), which costs you C's fees plus epsilon; then you create tx E that spends A with your desired feerate. So you're paying ~100kvB at the "backlog" rate here. 

With a 4kvB "related txs" limit, I think the worst case setup is just A is small (100vB) and is spent by B which is large (3900vB), so you have to pay at least 3.9kvB at the backlog rate plus epsilon to replace it with something smaller, but that's clearly already 25x better. Even if the backlog rate is not much smaller than the next block rate, you're only overpaying by at most 39x rather than perhaps 1000x here.

> I interpret this to mean that we should change both ancestor and descendant limits to 4000vB. 

Hmm, I think the worst case scenario here would still be bad:

 * A is spent by {B1,..,B20} (descendent count = 21 so less than 25)
 * B1 also spends C1, B2 also spends C2, etc; all via ephemeral outputs
 * A, Bn are 100vB each; Cn is 3800vB each

A has a descendent size of 2100vB; Bn has an ancestor size of 4000vB and a descendent size of 100vB; Cn has a descendent size of 3900vB. So the 4000vB restriction is okay.

But replacing A thus means Bn are invalid, which then means Cn are invalid due to the now unspent ephemeral outputs, which means you're replacing 78kvB in the mempool, not just 4kvB.

Reply:
@ajtowns Note that ephemeral outputs aren't proposed here, just the anc/desc limits.

Also note that V3 descendant count is limited to 2 (i.e. a V3 transaction can only have 1 descendant), so A can't be spent by {B1...B20}. But of course B1 could spend a bunch of things, up to {C1...C23}.

Good point that with ephemeral outputs replacing a child means we need to remove the parent as well. I'm experimenting with implementing it, and my idea was to require the parent (the one with the dust output) pay 0 fee so it can't stay in the mempool by itself. We'd still want to prevent replacing a large volume of transactions (i.e. include it in the RBF 100 limit), but users/devs don't have to worry about high costs of replacement since its fee is 0?

I'm thinking about ephemeral outputs but hoping to minimize the scope of this proposal. Correct me if I'm wrong, but it seems ephemeral outputs are not needed to make package relay work for LN penalty without pinning attacks. We could do ephemeral outputs as a next step - IIUC ephemeral outputs would be a "loosening" of policy rules so we could add it on to V3 rules later?

Reply:
> I believe this falls short in that the parent V3 transaction can be just south of 101kvB, even in ANYONECANPAY and SIGHASH_SINGLE like scenarios where additional inputs cannot be excluded by the presigning parties.

Can't we easily fix that by relaxing the 100kvB policy limit to be 104kvB for v3 transactions?

Reply:
> Can't we easily fix that by relaxing the 100kvB policy limit to be 104kvB for v3 transactions?

You would still have to replace 100kvB of griefing txns in general, so that doesn't sidestep the issue?

Reply:
That is true, it's more important to focus on the general issue taking the parents into account, you're right.